File: Algorithm_Note/Minimum_Spanning_Tree/Kruskals.java
//boj 1197
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.Scanner;

public class Kruskals {
	static Scanner sc;
	private int parent[];
	private int size;
	private long minCost=0;
	static PriorityQueue<Edge> pq;
	public Kruskals(int n) {
		this.size=n;
		this.parent=new int[size];
		Arrays.fill(parent,-1);
		int e=sc.nextInt();
		pq=new PriorityQueue<Edge>();
		for(int i=0; i<e; i++) 
			pq.add(new Edge(sc.nextInt()-1,sc.nextInt()-1,sc.nextInt()));
		kruskal();
	}
	public static void main(String[] args) {
		sc=new Scanner(System.in);
		int n=sc.nextInt();
		Kruskals x=new Kruskals(n);
	}

	public void weightedUnion(int i,int j){
		int x=collapsingFind(i);
		int y=collapsingFind(j);
		int newsize = parent[x] + parent[y];
		if (parent[x] >= parent[y]) {
			parent[x] = y;
			parent[y] = newsize;
		}
		else { 
			parent[y] = x;
			parent[x] = newsize;
		}
	}

	public int collapsingFind(int i){
		if(parent[i]<0)
			return i;
		return parent[i]=collapsingFind(parent[i]);
	}

	public void kruskal() {
		int mstSize=0;
		while(!pq.isEmpty()) {
			Edge x=pq.poll();
			if(collapsingFind(x.v)!=collapsingFind(x.w)) {
				weightedUnion(x.v,x.w);
				mstSize++;
				minCost+=x.weight;
				if(mstSize==size-1)
					break;
			}
		}
		System.out.println(minCost);
	}

	public static class Edge implements Comparable<Edge>{
		int v;
		int w;
		int weight;
		boolean selected;
		public Edge(int v,int w,int weight) {
			this.v=v;
			this.w=w;
			this.weight=weight;
		}
		@Override
		public int compareTo(Edge o) {
			if(this.weight-o.weight!=0)
				return this.weight-o.weight;
			else {
				if(this.v-o.v!=0)
					return this.v-o.v;
				else return this.w-o.w;
			}
		}
	}

}

File: Algorithm_Note/Minimum_Spanning_Tree/kruskals.cpp
//boj 1197
#include <stdio.h>
#include <string.h>
#include <algorithm>
#include <functional>
#include <queue>
using namespace std;

struct edge {
	int from, to, weight;
};
typedef struct edge Edge;

bool operator<(edge a,edge b) {
	return a.weight>b.weight;
}
priority_queue<edge> pq;
int parent[10001];

int CollapsingFind(int i) {
	if (parent[i] < 0)return i;
	return parent[i] = CollapsingFind(parent[i]);
}

void WeightedUnion(int i, int j) {
	int x = CollapsingFind(i);
	int y = CollapsingFind(j);
	if (parent[x] < parent[y]) {
		parent[x] += parent[y];
		parent[y] = x;
	}
	else {
		parent[y] += parent[x];
		parent[x] = y;
	}
}

int main() {
	memset(parent, -1, sizeof(parent));
	int V, E; scanf("%d %d", &V,&E);

	for (int i = 0; i < E; i++) {
		Edge* e = new Edge();
		scanf("%d %d %d", &e->from, &e->to, &e->weight);
		pq.push(*e);
	}

	int TreeSize = 0; int MinCost = 0;
	
	while (!pq.empty()) {
		Edge curr = pq.top(); pq.pop();
		if (CollapsingFind(curr.from) != CollapsingFind(curr.to))
			WeightedUnion(curr.from, curr.to),TreeSize++, MinCost += curr.weight;
		if (TreeSize == V - 1)break;
	}

	printf("%d", MinCost);
	
	return 0;
}